app/controllers/sessions_controller.rb

```diff
@@ -27,15 +27,31 @@ def create
   def twitter
     # raise request.env["omniauth.auth"].to_yaml
     auth = request.env["omniauth.auth"]
-    user = User.where(uid: auth["uid"]).first || User.from_twitter(auth)
+    if signed_in?
+      user = current_user
+      user.update_with_twitter(auth)
+      notice = "Your account is now linked with your Twitter account."
+      redirect_url = settings_url
+    else
+      user = User.where(uid: auth["uid"]).first || User.from_twitter(auth)
+      notice = "You have been logged in through Twitter."
+      redirect_url = root_url
+    end
     session[:user_id] = user.id
-    flash[:notice] = "You have been logged in through Twitter."
-    redirect_back_or root_url
+    flash[:notice] = notice
+    redirect_back_or redirect_url
   end
 
```

Comment:
Updated sessions#twitter to handle redirects from Twitter OAuth API when a user is already logged in. When a user is already logged in, the user was logged in using the "local" account.